File: leetcode/javaStuff/reshapeMatrix/ReshapeMatrix.java
public class ReshapeMatrix {
  public static void main(String[] args) {
    int[][] test = {{1,2,3}, {4,5,6}, {1,2,3}};

    int[] answer = extractElements(test);

    for(int x : answer) {
      System.out.println(x);
    }

  }


  // O(n) time and space
  // extract out all the elements of the original 2d array into single array

  // O(n) time
  // Go through that new array and allocate it accordingly to new 2d array
  // based on the new rows and cols from the input

  // make sure that input rows * cols is equal to ogArray's length * ogArray[0].length
  // or else return the original array


  // public int[][] reshapeMatrix(int[][] ogArray, int rows, int cols) {
  //   // base case if new array is impossible
  //   if(ogArray.length * ogArray[0].length != rows * cols) {
  //     return ogArray;
  //   }
  //
  //   int[] allElements = extractElements(ogArray);
  //
  // }

  public static int[] extractElements(int[][] array) {
    int[] extractedElements = new int[array.length * array[0].length];
    int counter = 0;

    for(int i = 0; i < array.length; i++) {
      for(int j = 0; j < array[i].length; j++) {
        extractedElements[counter] = array[i][j];
        counter++;
      }
    }

    return extractedElements;
  }
}

File: leetcode/reshape_matrix.js
// O(n) time
// O(n) space

function matrixReshape(nums, r, c) {
  let store = flatten(nums);
  let answer = createArray(r, c);
  let counter = 0;

  for(let i = 0; i < r; i++) {
    for(let j = 0; j < c; j++) {
      answer[i][j] = store[counter];
      counter++;
    }
  }

  return answer;
}

function createArray(r,c) {
  let arr = new Array(r);

  for(let i = 0; i < arr.length; i++) {
    arr[i] = new Array(c);
  }

  return arr;
}

function flatten(nums) {
  let store = [];

  for(let i = 0; i < nums.length; i++) {
    for(let j = 0; j < nums[0].length; j++) {
      store.push(nums[i][j]);
    }
  }

  return store;
}

let test = [[1,2],[4,5],[7,8]];
console.log(matrixReshape(test, 6, 1));
